game/game/damagecalculator.cpp

```diff
@@ -51,17 +51,44 @@ DamageCalculator::Val DamageCalculator::damageFall(Npc& npc, float speed) {
   }
 
 DamageCalculator::Val DamageCalculator::rangeDamage(Npc& nsrc, Npc& nother, const Bullet& b, const CollideMask bMsk) {
-  bool invinsible = !checkDamageMask(nsrc,nother,&b);
-  if(b.pathLength() > float(MaxBowRange) * b.hitChance() && b.hitChance()<1.f)
-    return Val(0,false,invinsible);
+  float dist       = b.pathLength();
+  bool  noHit      = dist>float(MaxMagRange);
+  bool  invincible = !checkDamageMask(nsrc,nother,&b);
+  auto  dmg        = b.damage();
+
+  if(!b.isSpell()) {
+    auto& script    = nsrc.world().script();
+    int   hitChance = int(script.rand(100));
+    int   hitCh     = 0;
+    bool  g2        = Gothic::inst().version().game==2;
+    float critRange = g2 ? CritBowRange : (CritBowRange + 500);
+    float skill     = b.hitChanceVal();
+
+    if(dist<critRange)
+      hitCh = int((skill - 100.f) / critRange * dist + 100.f + 0.5f); else
```

Comment:
Waaaaait... this is `((critRange-dist)/critRange)*skill*100.f + 0.5f`?

Also measuring probability in 0..100 rage is awkward. It makes way more sense to use 0..1, as in probability theory, and cast it to integer scale at very end, than you'll get:
`((critRange-dist)/critRange)*skill`

Reply:
If `dist=critRange` both formulas give `0` instead of `skill`. I just used 2 point formula for linear functions. I thought it's better to let it as is and let compiler optimize.

I used 0..100 range because of the comparison with the random generated number which is in this range in `swordDamage`. But 0..1 range is also ok and avoids rounding.

```C++
    float hitChance = float(script.rand(100))/100.f;
    if(dist<refRange)
      hitCh = (skill - 1.f) / refRange * dist + 1.f; else
      hitCh = skill / (refRange - float(MaxBowRange) * (dist - float(MaxBowRange)));
    noHit = (dist>float(MaxBowRange) || hitCh<=hitChance);
```